pkg/services/user/password.go

```diff
@@ -0,0 +1,57 @@
+package user
+
+import (
+	"unicode"
+
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/util/errutil"
+)
+
+var (
+	ErrPasswordTooShort       = errutil.NewBase(errutil.StatusBadRequest, "password-policy-too-short", errutil.WithPublicMessage("New password is too short"))
+	ErrPasswordPolicyInfringe = errutil.NewBase(errutil.StatusBadRequest, "password-policy-infringe", errutil.WithPublicMessage("New password doesn't comply with the password policy"))
+	MinPasswordLength         = 12
+)
+
+/*
+Static function for password validation
+*/
+func ValidatePassword(newPassword Password, config *setting.Cfg) error {
```

Comment:
I think we should get rid of `Password` and just use string. Then move just recreate `IsWeek` but as a non exported function here.

Reply:
I think we don't need to fall into [Primitive Obsession](https://refactoring.guru/smells/primitive-obsession).

I think we just need to change

```go
userPassword := user.Password(newPassword)
	if err := user.ValidatePassword(userPassword, runner.Cfg); err != nil {
		return fmt.Errorf("new password is too short")
	}
```

to 

```go
userPassword, err := user.NewPassword(newPassword)
...
```

Make validation part of the type and not an afterthought

Reply:
this follows more a pattern of Parsing, not validating

https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

Reply:
Yep that is also a good suggestion

Reply:
Then you ideally would also update the type in models to not use string but password type instead

Reply:
I migrated all he password properties from string type to user.Password type. I'm working on the tests